enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java

```diff
@@ -54,56 +70,263 @@
+                if ( marked )
+                {
+                    terminatedCount++;
+                }
+            }
+        }
+        return Stream.of( new TransactionTerminationResult( username, terminatedCount ) );
+    }
+
+    private Stream<ConnectionResult> terminateConnectionsForValidUser( String username )
+    {
+        Long killCount = 0L;
+        for ( ManagedBoltStateMachine connection : getBoltConnectionTracker().getActiveConnections( username ) )
+        {
+            connection.terminate();
+            killCount += 1;
+        }
+        return Stream.of( new ConnectionResult( username, killCount ) );
+    }
+
+    private Set<KernelTransactionHandle> getActiveTransactions()
+    {
+        return graph.getDependencyResolver().resolveDependency( KernelTransactions.class ).activeTransactions();
+    }
+
+    private BoltConnectionTracker getBoltConnectionTracker()
+    {
+        return graph.getDependencyResolver().resolveDependency( BoltConnectionTracker.class );
+    }
+
+    private Stream<TransactionResult> countTransactionByUsername( Stream<String> usernames )
+    {
+        return usernames
+            .collect( Collectors.groupingBy( Function.identity(), Collectors.counting() ) )
+            .entrySet()
+            .stream()
+            .map( entry -> new TransactionResult( entry.getKey(), entry.getValue() )
+        );
+    }
+
+    private Stream<ConnectionResult> countConnectionsByUsername( Stream<String> usernames )
+    {
+        return usernames
+            .collect( Collectors.groupingBy( Function.identity(), Collectors.counting() ) )
+            .entrySet()
+            .stream()
+            .map( entry -> new ConnectionResult( entry.getKey(), entry.getValue() )
+        );
+    }
+
+    private boolean isAdminEnterpriseAuthSubject()
+    {
+        if ( authSubject instanceof EnterpriseAuthSubject )
+        {
+            EnterpriseAuthSubject enterpriseAuthSubject = (EnterpriseAuthSubject) authSubject;
+            return enterpriseAuthSubject.isAdmin();
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    private EnterpriseAuthSubject ensureAdminEnterpriseAuthSubject()
     {
-        EnterpriseAuthSubject enterpriseAuthSubject = (EnterpriseAuthSubject) authSubject;
-        return enterpriseAuthSubject.isAdmin();
+        EnterpriseAuthSubject enterpriseAuthSubject = EnterpriseAuthSubject.castOrFail( authSubject );
+        if ( !enterpriseAuthSubject.isAdmin() )
+        {
+            throw new AuthorizationViolationException( PERMISSION_DENIED );
+        }
+        return enterpriseAuthSubject;
+    }
+
+    private EnterpriseAuthSubject ensureSelfOrAdminEnterpriseAuthSubject( String username )
+            throws InvalidArgumentsException
+    {
+        EnterpriseAuthSubject subject = EnterpriseAuthSubject.castOrFail( authSubject );
+        subject.ensureUserExistsWithName( username );
```

Comment:
Shouldn't this check be after the check for admin or self? Otherwise we could leak information about what users are available to a non admin user.